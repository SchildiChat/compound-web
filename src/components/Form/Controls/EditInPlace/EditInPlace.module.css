/*
Copyright 2024 New Vector Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

.container {
  min-block-size: 124px;
}

.label {
  font-size: 15px;
  font-weight: 500;
  margin-block-end: 4px;
}

.controls {
  display: flex;
  gap: 15px;
}

.button-group {
  display: none;
  position: relative;
  inset-block-start: 5px;
  grid-template-columns: 1fr 1fr;
  gap: 7px;
}

.button {
  display: flex;
  inline-size: 36px;
  block-size: 36px;
  border: 1px solid var(--cpd-color-border-interactive-secondary);
  border-radius: 20px;
  background-color: transparent;
  text-align: center;
  cursor: pointer;
}

.primary-button {
  background-color: var(--cpd-color-bg-action-primary-rest);
  border-color: var(--cpd-color-bg-action-primary-rest);

  svg {
    color: var(--cpd-color-icon-on-solid-primary);
  }
}

.primary-button-disabled {
  background-color: var(--cpd-color-bg-subtle-primary);
  border-color: var(--cpd-color-bg-subtle-primary);

  svg {
    color: var(--cpd-color-icon-disabled);
  }
}

.button svg {
  inline-size: 24px;
  block-size: 24px;
  margin: auto;
}

.control {
  inline-size: 100%;
}

.container-show-buttons .button-group,
.container:focus-within .button-group {
  display: inline-grid;
}

.container-error .control {
  border-color: var(--cpd-color-border-critical-primary);
}

.container-error .label {
  color: var(--cpd-color-text-critical-primary);
}

.caption-line {
  margin-block-start: var(--cpd-space-2x);
  display: flex;
  align-items: center;
  gap: var(--cpd-space-2x);
}

.caption-icon {
  display: inline-block;
  inline-size: 20px;
  block-size: 20px;
}

.caption-icon-error {
  color: var(--cpd-color-icon-critical-primary);
}

.caption-icon-saved {
  background-color: var(--cpd-color-icon-success-primary);

  /*
   * We are trying to match the size of the error icon which
   * is displayed 20px big but has 2px of border in an SVG
   * with a viewBox of 24x24...
   */
  inline-size: calc(20px * (20 / 24));
  block-size: calc(20px * (20 / 24));
  margin: calc(2px * (20 / 24));
  border-radius: 20px;
  text-align: center;

  svg {
    color: var(--cpd-color-icon-on-solid-primary);
    position: relative;
    inset-block-start: -1px;
  }
}

.caption-text {
  font-size: var(--cpd-font-body-sm-medium);
  font-weight: var(--cpd-font-weight-medium);
}

.caption-text-error {
  color: var(--cpd-color-text-critical-primary);
}

.caption-text-saved {
  color: var(--cpd-color-text-success-primary);
}

.caption-text-saving {
  color: var(--cpd-color-text-secondary);
}
