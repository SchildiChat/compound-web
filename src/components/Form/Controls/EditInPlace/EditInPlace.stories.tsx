/*
Copyright 2024 New Vector Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React from "react";

import { EditInPlace } from "./";
import { Meta, StoryObj } from "@storybook/react";

type Props = { invalid?: boolean } & React.ComponentProps<typeof EditInPlace>;

export default {
  title: "Form/Controls/EditInPlace",
  component: EditInPlace,
  tags: ["autodocs"],
  parameters: {
    controls: {
      include: [
        "onChange",
        "onSave",
        "onCancel",
        "value",
        "valueIsChanged",
        "error",
        "savedLabel",
        "saveButtonLabel",
        "cancelButtonLabel",
      ],
    },
  },
  argTypes: {
    label: {
      type: "string",
    },
    value: {
      type: "string",
    },
    valueIsChanged: {
      type: "boolean",
    },
    onChange: {
      action: "changed",
    },
    onSave: {
      action: "saved",
    },
    onCancel: {
      action: "cancelled",
    },
    error: {
      type: "string",
    },
    savedLabel: {
      type: "string",
    },
    saveButtonLabel: {
      type: "string",
    },
    cancelButtonLabel: {
      type: "string",
    },
  },
  render: ({ ...restArgs }) => <EditInPlace {...restArgs} />,
  args: {
    label: "Label",
    value: "",
    saveButtonLabel: "Save",
    cancelButtonLabel: "Cancel",
  },
} satisfies Meta<Props>;

type Story = StoryObj<Props>;

export const Empty: Story = {
  parameters: {
    design: {
      type: "figma",
      url: "https://www.figma.com/file/rTaQE2nIUSLav4Tg3nozq7/Compound-Web-Components?type=design&node-id=4335-2016&mode=design&t=BvxRca0YDRaq20IR-0",
    },
  },
};
