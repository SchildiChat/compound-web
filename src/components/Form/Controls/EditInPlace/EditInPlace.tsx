/*
Copyright 2024 New Vector Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import classnames from "classnames";
import React, {
  forwardRef,
  useCallback,
  useRef,
  useState,
  useEffect,
} from "react";
import { Submit } from "@radix-ui/react-form";
import CheckIcon from "@vector-im/compound-design-tokens/icons/check.svg";
import CancelIcon from "@vector-im/compound-design-tokens/icons/close.svg";

import styles from "./EditInPlace.module.css";
import { TextControl } from "../Text";

import {
  ErrorMessage,
  Field,
  Label,
  LoadingMessage,
  Root,
  SuccessMessage,
} from "../..";
import { Tooltip } from "../../../..";

type Props = {
  /**
   * The label for the control
   */
  label: string;

  /**
   * The CSS class name.
   */
  className?: string;

  /**
   * Callback for when the user confirms the change
   */
  onSave?: (e: React.FormEvent) => Promise<void> | void;

  /**
   * Callback for when the user wishes to cancel the change
   */
  onCancel?: (e: React.FormEvent) => void;

  /**
   * onInput event handler on the text control
   */
  onInput?: (e: React.ChangeEvent<HTMLInputElement>) => void;

  /**
   * Error message to be displayed below the box. If supplied, will disable the
   * save button.
   */
  error?: string;

  /**
   * Label to be displayed by the green check at the bottom. Will only be displayed
   * for 2 seconds after the onSave callback promise resolves successfully.
   */
  savedLabel?: string;

  /**
   * The label for the save button
   */
  saveButtonLabel: string;

  /**
   * The label for the 'in progress' saving caption
   */
  savingLabel: string;

  /**
   * The label for the cancel button
   */
  cancelButtonLabel: string;

  /**
   * If true, disabled the entire component to disallow editing.
   */
  disabled?: boolean;
} & React.ComponentProps<typeof TextControl>;

/**
 * A text box with save/cancel buttons that appear when the field is active.
 * Since thios control has its own 'save' button, it should *not* appear as part
 * of a larger form: it exists as its own form that submits separately.
 */
export const EditInPlace = forwardRef<HTMLInputElement, Props>(
  function EditInPlace(
    {
      className,
      label,
      onSave,
      onCancel,
      saveButtonLabel,
      cancelButtonLabel,
      error,
      savedLabel,
      savingLabel,
      disabled,
      ...props
    },
    ref,
  ) {
    const [showSaved, setShowSaved] = useState(false);
    const [saving, setSaving] = useState(false);
    const [dirty, setDirty] = useState(false);

    const classes = classnames(styles.container, className, {
      [styles["is-dirty"]]: dirty,
    });

    const hideTimer = useRef<ReturnType<typeof setTimeout>>();

    const saveButtonRef = useRef<HTMLButtonElement | null>(null);
    const cancelButtonRef = useRef<HTMLButtonElement | null>(null);

    useEffect(() => {
      return () => {
        // Clear any timers that may have been set when the component gets unmounted
        if (hideTimer.current) clearTimeout(hideTimer.current);
        hideTimer.current = undefined;
      };
    }, []);

    const onInput = useCallback(
      (e: React.ChangeEvent<HTMLInputElement>) => {
        setDirty(true);
        props.onInput?.(e);
      },
      [setDirty],
    );

    const onFormSubmit = useCallback(
      async (e: React.FormEvent) => {
        e.preventDefault();
        try {
          setShowSaved(false);
          setSaving(true);
          saveButtonRef.current?.blur();
          await onSave?.(e);
          setDirty(false);
          setShowSaved(true);

          if (hideTimer.current) clearTimeout(hideTimer.current);
          hideTimer.current = setTimeout(() => {
            setShowSaved(false);
            hideTimer.current = undefined;
          }, 2000);
        } catch (e) {
          // We don't really need to do anything here, we just don't want to display the
          // 'saved' label, obviously. The user of the component can update the error to
          // show what failed.
        } finally {
          setSaving(false);
        }
      },
      [setShowSaved, onSave, hideTimer],
    );

    const onFormReset = useCallback(
      (e: React.FormEvent) => {
        cancelButtonRef.current?.blur();
        onCancel?.(e);
        setDirty(false);
      },
      [cancelButtonRef, onCancel],
    );

    return (
      <Root className={classes} onSubmit={onFormSubmit} onReset={onFormReset}>
        <Field name="input" serverInvalid={Boolean(error)}>
          <Label>{label}</Label>
          <div className={styles.controls}>
            <TextControl
              ref={ref}
              {...props}
              onInput={onInput}
              disabled={disabled || saving}
            />
            <div className={styles["button-group"]}>
              <Tooltip label={saveButtonLabel}>
                <Submit asChild>
                  <button
                    type="submit"
                    className={classnames(
                      styles.button,
                      styles["primary-button"],
                      {
                        [styles["primary-button-disabled"]]: !dirty,
                      },
                    )}
                    ref={saveButtonRef}
                    aria-label={saveButtonLabel}
                    disabled={!dirty || saving}
                  >
                    <CheckIcon />
                  </button>
                </Submit>
              </Tooltip>

              <Tooltip label={cancelButtonLabel}>
                <button
                  type="reset"
                  role="button"
                  ref={cancelButtonRef}
                  className={styles.button}
                  aria-label={cancelButtonLabel}
                  disabled={saving}
                >
                  <CancelIcon />
                </button>
              </Tooltip>
            </div>
          </div>
          {error && <ErrorMessage>{error}</ErrorMessage>}
          {saving && <LoadingMessage>{savingLabel}</LoadingMessage>}
          {savedLabel && showSaved && (
            <SuccessMessage>{savedLabel}</SuccessMessage>
          )}
        </Field>
      </Root>
    );
  },
);
