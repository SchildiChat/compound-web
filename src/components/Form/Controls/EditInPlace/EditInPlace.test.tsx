/*
Copyright 2024 New Vector Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import { describe, it, expect, vi, afterEach } from "vitest";
import React from "react";
import { render } from "@testing-library/react";

import { userEvent } from "@storybook/test";
import { act } from "react-dom/test-utils";
import { TooltipProvider } from "@radix-ui/react-tooltip";
import { EditInPlace } from "./EditInPlace";

type EditInPlaceTestProps = {
  error?: string;
  defaultValue?: string;
  onChange?: (e: React.ChangeEvent<HTMLInputElement>) => void;
  onSave?: () => Promise<void>;
  onCancel?: () => void;
  disabled?: boolean;
};

describe("EditInPlace", () => {
  const EditInPlaceTest = (props: EditInPlaceTestProps) => (
    <TooltipProvider>
      <EditInPlace
        label="Edit Me"
        defaultValue={props.defaultValue}
        error={props.error}
        onChange={props.onChange ?? (() => {})}
        onSave={props.onSave ?? (() => Promise.resolve())}
        onCancel={props.onCancel ?? (() => {})}
        saveButtonLabel="Save"
        cancelButtonLabel="Cancel"
        savedLabel="Saved"
        savingLabel="Saving..."
        disabled={props.disabled}
      />
    </TooltipProvider>
  );

  afterEach(() => {
    vi.useRealTimers();
  });

  it("renders", () => {
    const { asFragment } = render(<EditInPlaceTest />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("calls onChange when text edited", async () => {
    let value;
    const onChange = vi
      .fn()
      .mockImplementation((e: React.ChangeEvent<HTMLInputElement>) => {
        value = e.target.value;
      });

    const { getByRole } = render(
      <EditInPlaceTest onChange={onChange} defaultValue="Edit this text" />,
    );

    await act(async () => {
      const input = getByRole("textbox");
      await userEvent.type(input, "!");
    });

    expect(onChange).toHaveBeenCalled();
    expect(value).toEqual("Edit this text!");
  });

  it("field is valid if no error specified", () => {
    const { getByRole } = render(<EditInPlaceTest />);

    const input = getByRole("textbox");
    expect(input).toBeValid();
  });

  it("renders error icon and text if error provided", () => {
    const { asFragment, getByRole, getByText } = render(
      <EditInPlaceTest error="Missing Left Falangey" />,
    );
    expect(asFragment()).toMatchSnapshot();

    const input = getByRole("textbox");
    expect(input).toBeInvalid();
    expect(input).toHaveAttribute("aria-describedby");

    const errorText = getByText("Missing Left Falangey");
    expect(errorText).toBeInTheDocument();

    expect(errorText.id).toEqual(input.getAttribute("aria-describedby"));
  });

  it("enables save button once we entered something", async () => {
    const { getByRole } = render(<EditInPlaceTest />);

    expect(getByRole("button", { name: "Save" })).toBeDisabled();

    await act(async () => {
      await userEvent.type(getByRole("textbox"), "Changed");
    });

    expect(getByRole("button", { name: "Save" })).toBeEnabled();
  });

  it("calls save callback on save button click", async () => {
    const onSave = vi.fn();
    const { getByRole } = render(<EditInPlaceTest onSave={onSave} />);

    await act(async () => {
      await userEvent.type(getByRole("textbox"), "Changed");
      await userEvent.click(getByRole("button", { name: "Save" }));
    });

    expect(onSave).toHaveBeenCalled();
  });

  it("calls save callback if enter pressed in the text box", async () => {
    const onSave = vi.fn();
    const { getByRole } = render(<EditInPlaceTest onSave={onSave} />);

    await act(async () => {
      await userEvent.type(getByRole("textbox"), "Changed");
      await userEvent.type(getByRole("textbox"), "{enter}");
    });

    expect(onSave).toHaveBeenCalled();
  });

  it("calls onCancel when cancel button pressed", async () => {
    const onCancel = vi.fn();
    const { getByRole } = render(<EditInPlaceTest onCancel={onCancel} />);

    await act(async () => {
      await userEvent.type(getByRole("textbox"), "Changed");
      await userEvent.click(getByRole("button", { name: "Cancel" }));
    });

    expect(onCancel).toHaveBeenCalled();
  });

  it("resets the form when cancel button pressed", async () => {
    const onCancel = vi.fn();
    const { getByRole } = render(
      <EditInPlaceTest onCancel={onCancel} defaultValue="Initial" />,
    );

    const input = getByRole("textbox") as HTMLInputElement;

    await act(async () => {
      await userEvent.clear(input);
      await userEvent.type(input, "Changed");
    });

    expect(input.value).toBe("Changed");

    await act(async () => {
      await userEvent.click(getByRole("button", { name: "Cancel" }));
    });

    expect(input.value).toBe("Initial");

    expect(onCancel).toHaveBeenCalled();
  });

  it("unfocuses the input when cancel button pressed", async () => {
    const { getByRole } = render(<EditInPlaceTest />);

    const input = getByRole("textbox");
    await act(async () => {
      await userEvent.type(input, "Changed");
      await userEvent.click(getByRole("button", { name: "Cancel" }));
    });

    expect(document.activeElement).not.toEqual(input);
  });

  it("unfocuses the input when the save callback promise resolves", async () => {
    const { getByRole } = render(<EditInPlaceTest />);

    const input = getByRole("textbox");
    await act(async () => {
      await userEvent.type(input, "Changed");
      await userEvent.click(getByRole("button", { name: "Save" }));
    });

    expect(document.activeElement).not.toEqual(input);
  });

  it("displays saved label for 2 seconds after save", async () => {
    vi.useFakeTimers();
    const user = userEvent.setup({ advanceTimers: vi.advanceTimersByTime });

    const { getByRole, getByText, queryByText, asFragment } = render(
      <EditInPlaceTest />,
    );

    expect(asFragment()).toMatchSnapshot();
    expect(queryByText("Saved")).not.toBeInTheDocument();

    await act(async () => {
      await user.type(getByRole("textbox"), "Changed");
      await user.click(getByRole("button", { name: "Save" }));
    });

    expect(asFragment()).toMatchSnapshot();
    expect(getByText("Saved")).toBeInTheDocument();

    act(() => {
      vi.advanceTimersByTime(1900);
    });

    expect(queryByText("Saved")).toBeInTheDocument();

    act(() => {
      vi.advanceTimersByTime(200);
    });

    expect(queryByText("Saved")).not.toBeInTheDocument();
  });

  it("does not call onSave if cancel pressed", async () => {
    const onSave = vi.fn();
    const { getByRole } = render(<EditInPlaceTest onSave={onSave} />);

    await act(async () => {
      await userEvent.type(getByRole("textbox"), "Changed");
      await userEvent.click(getByRole("button", { name: "Cancel" }));
    });

    expect(onSave).not.toHaveBeenCalled();
  });

  it("disables control when disabled", () => {
    const { getByRole, asFragment } = render(<EditInPlaceTest disabled />);
    expect(asFragment()).toMatchSnapshot();

    const input = getByRole("textbox");
    expect(input).toBeDisabled();
  });
});
