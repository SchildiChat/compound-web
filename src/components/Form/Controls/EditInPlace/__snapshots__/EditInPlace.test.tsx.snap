// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`EditInPlace > renders 1`] = `
<DocumentFragment>
  <form
    class="_root_dc87d2 _container_980156"
  >
    <div
      class="_field_dc87d2"
    >
      <label
        class="_label_dc87d2"
        for="radix-:r0:"
      >
        Edit Me
      </label>
      <div
        class="_controls_980156"
      >
        <input
          class="_control_a839aa"
          id="radix-:r0:"
          name="input"
          title=""
          value=""
        />
        <div
          class="_button-group_980156"
        >
          <button
            aria-label="Save"
            class="_button_980156 _primary-button_980156"
            type="submit"
          >
            <svg
              class="cpd-icon"
              fill="currentColor"
              height="1em"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M9.55 17.575c-.133 0-.258-.02-.375-.063a.878.878 0 0 1-.325-.212L4.55 13c-.183-.183-.27-.42-.263-.713.009-.291.105-.529.288-.712a.948.948 0 0 1 .7-.275.95.95 0 0 1 .7.275L9.55 15.15l8.475-8.475c.183-.183.42-.275.712-.275s.53.092.713.275c.183.183.275.42.275.712s-.092.53-.275.713l-9.2 9.2c-.1.1-.208.17-.325.212a1.106 1.106 0 0 1-.375.063Z"
              />
            </svg>
          </button>
          <button
            aria-label="Cancel"
            class="_button_980156"
            role="button"
            type="reset"
          >
            <svg
              class="cpd-icon"
              fill="currentColor"
              height="1em"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M6.293 6.293a1 1 0 0 1 1.414 0L12 10.586l4.293-4.293a1 1 0 1 1 1.414 1.414L13.414 12l4.293 4.293a1 1 0 0 1-1.414 1.414L12 13.414l-4.293 4.293a1 1 0 0 1-1.414-1.414L10.586 12 6.293 7.707a1 1 0 0 1 0-1.414Z"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
  </form>
</DocumentFragment>
`;

exports[`EditInPlace > renders error icon and text if error provided 1`] = `
<DocumentFragment>
  <form
    class="_root_dc87d2 _container_980156"
  >
    <div
      class="_field_dc87d2"
      data-invalid="true"
    >
      <label
        class="_label_dc87d2"
        data-invalid="true"
        for="radix-:rr:"
      >
        Edit Me
      </label>
      <div
        class="_controls_980156"
      >
        <input
          aria-describedby="radix-:r14:"
          aria-invalid="true"
          class="_control_a839aa"
          data-invalid="true"
          id="radix-:rr:"
          name="input"
          title=""
          value=""
        />
        <div
          class="_button-group_980156"
        >
          <button
            aria-label="Save"
            class="_button_980156 _primary-button_980156"
            type="submit"
          >
            <svg
              class="cpd-icon"
              fill="currentColor"
              height="1em"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M9.55 17.575c-.133 0-.258-.02-.375-.063a.878.878 0 0 1-.325-.212L4.55 13c-.183-.183-.27-.42-.263-.713.009-.291.105-.529.288-.712a.948.948 0 0 1 .7-.275.95.95 0 0 1 .7.275L9.55 15.15l8.475-8.475c.183-.183.42-.275.712-.275s.53.092.713.275c.183.183.275.42.275.712s-.092.53-.275.713l-9.2 9.2c-.1.1-.208.17-.325.212a1.106 1.106 0 0 1-.375.063Z"
              />
            </svg>
          </button>
          <button
            aria-label="Cancel"
            class="_button_980156"
            role="button"
            type="reset"
          >
            <svg
              class="cpd-icon"
              fill="currentColor"
              height="1em"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M6.293 6.293a1 1 0 0 1 1.414 0L12 10.586l4.293-4.293a1 1 0 1 1 1.414 1.414L13.414 12l4.293 4.293a1 1 0 0 1-1.414 1.414L12 13.414l-4.293 4.293a1 1 0 0 1-1.414-1.414L10.586 12 6.293 7.707a1 1 0 0 1 0-1.414Z"
              />
            </svg>
          </button>
        </div>
      </div>
      <span
        class="_message_dc87d2 _error-message_dc87d2"
        id="radix-:r14:"
      >
        <svg
          class="cpd-icon"
          fill="currentColor"
          height="1em"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12 17a.97.97 0 0 0 .713-.288A.968.968 0 0 0 13 16a.968.968 0 0 0-.287-.713A.968.968 0 0 0 12 15a.968.968 0 0 0-.713.287A.968.968 0 0 0 11 16c0 .283.096.52.287.712.192.192.43.288.713.288Zm0-4c.283 0 .52-.096.713-.287A.968.968 0 0 0 13 12V8a.967.967 0 0 0-.287-.713A.968.968 0 0 0 12 7a.968.968 0 0 0-.713.287A.967.967 0 0 0 11 8v4c0 .283.096.52.287.713.192.191.43.287.713.287Zm0 9a9.738 9.738 0 0 1-3.9-.788 10.099 10.099 0 0 1-3.175-2.137c-.9-.9-1.612-1.958-2.137-3.175A9.738 9.738 0 0 1 2 12a9.74 9.74 0 0 1 .788-3.9 10.099 10.099 0 0 1 2.137-3.175c.9-.9 1.958-1.612 3.175-2.137A9.738 9.738 0 0 1 12 2a9.74 9.74 0 0 1 3.9.788 10.098 10.098 0 0 1 3.175 2.137c.9.9 1.613 1.958 2.137 3.175A9.738 9.738 0 0 1 22 12a9.738 9.738 0 0 1-.788 3.9 10.098 10.098 0 0 1-2.137 3.175c-.9.9-1.958 1.613-3.175 2.137A9.738 9.738 0 0 1 12 22Z"
          />
        </svg>
        Missing Left Falangey
      </span>
    </div>
  </form>
</DocumentFragment>
`;

exports[`EditInPlace > should disable save button if told to 1`] = `
<DocumentFragment>
  <form
    class="_root_dc87d2 _container_980156"
  >
    <div
      class="_field_dc87d2"
    >
      <label
        class="_label_dc87d2"
        for="radix-:r15:"
      >
        Edit Me
      </label>
      <div
        class="_controls_980156"
      >
        <input
          class="_control_a839aa"
          id="radix-:r15:"
          name="input"
          title=""
          value=""
        />
        <div
          class="_button-group_980156"
        >
          <button
            aria-label="Save"
            class="_button_980156 _primary-button_980156 _primary-button-disabled_980156"
            disabled=""
            type="submit"
          >
            <svg
              class="cpd-icon"
              fill="currentColor"
              height="1em"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M9.55 17.575c-.133 0-.258-.02-.375-.063a.878.878 0 0 1-.325-.212L4.55 13c-.183-.183-.27-.42-.263-.713.009-.291.105-.529.288-.712a.948.948 0 0 1 .7-.275.95.95 0 0 1 .7.275L9.55 15.15l8.475-8.475c.183-.183.42-.275.712-.275s.53.092.713.275c.183.183.275.42.275.712s-.092.53-.275.713l-9.2 9.2c-.1.1-.208.17-.325.212a1.106 1.106 0 0 1-.375.063Z"
              />
            </svg>
          </button>
          <button
            aria-label="Cancel"
            class="_button_980156"
            role="button"
            type="reset"
          >
            <svg
              class="cpd-icon"
              fill="currentColor"
              height="1em"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M6.293 6.293a1 1 0 0 1 1.414 0L12 10.586l4.293-4.293a1 1 0 1 1 1.414 1.414L13.414 12l4.293 4.293a1 1 0 0 1-1.414 1.414L12 13.414l-4.293 4.293a1 1 0 0 1-1.414-1.414L10.586 12 6.293 7.707a1 1 0 0 1 0-1.414Z"
              />
            </svg>
          </button>
        </div>
      </div>
      <span
        class="_caption-line_980156 _caption-text_980156 _caption-text-help_980156"
      >
        Help message
      </span>
    </div>
  </form>
</DocumentFragment>
`;
